Fixes incorrect product redirect after edit #13930

PADMASRIAMBATI opened the pull request:
##Proposed Changes
Fixes #13930

Modified the product editing success handler to use the product details route for redirection instead of the product listing route.

Confirmed that the redirect logic utilizes the product's unique identifier (ID) to correctly construct the target URL, ensuring the user lands exactly on the details page of the item they just edited.

This restores the expected user flow: Edit → Save → Product Details Page.

Additional Context
This is a user experience (UX) fix that prevents the user from being unnecessarily redirected to the product list and having to click back into the details page manually after a successful edit.

Tagging: @ohcnetwork/care-fe-code-reviewers
## Merge Checklist

- [x] Add specs that demonstrate the bug or test the new feature.
- [x] Update [product documentation](https://docs.ohc.network).
- [x] Ensure that UI text is placed in I18n files.
- [x] Prepare a screenshot or demo video for the changelog entry and attach it to the issue.
- [x] Request peer reviews.
- [x] Complete QA on mobile devices.
- [x] Complete QA on desktop devices.


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

* **New Features**
  * After updating a product, you’re automatically redirected to that product’s detail page, providing immediate confirmation and quicker access to follow-up actions.
  * URLs now reflect the specific product (e.g., /facility/{facilityId}/settings/product/{productId}) after an update, improving clarity and shareability.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

PADMASRIAMBATI committed:
Fix: Redirects user to product details after successful edit

tasnimshaji commented:
Looks good to me @nihal467

PADMASRIAMBATI committed:
Merge branch 'ohcnetwork:develop' into issues/13930/redirect-after-edit

PADMASRIAMBATI committed:
Chore(product): Final lint and formatting cleanup on ProductForm.tsx

PADMASRIAMBATI commented:
"Hello @ohcnetwork/care-fe-code-reviewers,

The code is now finalized and addresses all points from the review thread:

Redirect Fix: The user is correctly redirected to the specific product details page after a successful edit.

Consistency: The onSuccess handlers for both create and update mutations now correctly use the flexible onSuccess prop, as suggested in the CodeRabbit review.

Code Quality: All linting, formatting, and TypeScript nits have been resolved.

The latest commit contains these fixes. The PR is now complete and ready for final approval and merge. Thank you!"

NikhilA8606 reviewed (commented):
Similar issue found in the Product Knowledge form, solve that too in this PR

PADMASRIAMBATI committed:
Merge branch 'ohcnetwork:develop' into issues/13930/redirect-after-edit

PADMASRIAMBATI committed:
Fix(product-knowledge): Resolve redirect issue after edit and unify onSuccess handlers

PADMASRIAMBATI committed:
Fix(product-knowledge): Resolve redirect issue after edit and unify onSuccess handlers

PADMASRIAMBATI committed:
Fix(product-knowledge): Resolve redirect issue after edit and unify onSuccess handlers

PADMASRIAMBATI committed:
Fix(forms): Finalize redirect logic in Product and ProductKnowledge forms, including onCancel fix.

PADMASRIAMBATI committed:
Fix(forms): Finalize redirect logic in Product and ProductKnowledge forms, including onCancel fix.

PADMASRIAMBATI committed:
Fix(forms): Finalize redirect logic in Product and ProductKnowledge forms; addresses onCancel consistency. #13930

PADMASRIAMBATI committed:
Merge branch 'ohcnetwork:develop' into issues/13930/redirect-after-edit

PADMASRIAMBATI committed:
Chore(docs): Add missing docstrings to ProductForm to fix coverage warning.

PADMASRIAMBATI committed:
final update

abhimanyurajeesh commented on src/pages/Facility/settings/productKnowledge/ProductKnowledgeForm.tsx:
```diff
@@ -1004,8 +1001,8 @@ function ProductKnowledgeFormContent({
-                      : `/product_knowledge/categories/${categorySlug}`
+                      ? `/facility/${facilityId}/settings/product_knowledge/${slug}`
+                      : `/facility/${facilityId}/settings/product_knowledge/categories/${categorySlug}`
                   }
```
?

PADMASRIAMBATI replied:
Yes, this conditional logic is correct and intentional for UX.

It handles the "Cancel" button on the form by redirecting the user to the most logical page based on context:
    1. If Editing (isEditMode is true): Returns to the existing product's Details Page (.../${slug}).
    2. If Creating (isEditMode is false): Returns to the Category Listing Page (.../categories/${categorySlug}).
This pattern prevents the user from being unintentionally dropped at a generic list page when canceling an edit.

PADMASRIAMBATI committed:
Fix(redirects): Resolve incorrect post-edit redirection for Product, ProductKnowledge, and TokenCategory forms. Adds comprehensive E2E tests for verification.

abhimanyurajeesh reviewed (dismissed):
lets ignore ProductKnowledgeForm as its taken care in another PR

PADMASRIAMBATI committed:
Fix(redirects): Remove overlapping ProductKnowledge files as requested by reviewer, retaining fixes for Product and TokenCategory.

PADMASRIAMBATI committed:
Delete tests/product-knowledge-forms.spec.ts

PADMASRIAMBATI committed:
Merge branch 'ohcnetwork:develop' into issues/13930/redirect-after-edit

bodhish commented on tests/product-forms.spec.ts:
The location for the file doesn't look right

bodhish commented on src/pages/Facility/settings/tokenCategory/TokenCategoryForm.tsx:
```diff
@@ -114,8 +114,18 @@ export function TokenCategoryFormContent({
   containerClassName,
-  onSuccess = () => navigate(`/facility/${facilityId}/settings/token_category`),
-  onCancel = () => navigate(`/facility/${facilityId}/settings/token_category`),
+  onSuccess = (tokenCategory) =>
```
This looks like an unrelated change!

PADMASRIAMBATI committed:
Simplify navigation on success and cancel actions

PADMASRIAMBATI committed:
Merge branch 'ohcnetwork:develop' into issues/13930/redirect-after-edit

PADMASRIAMBATI committed:
Create product-forms.spec.ts

PADMASRIAMBATI committed:
Delete tests/product-forms.spec.ts

PADMASRIAMBATI committed:
Delete tests/e2e directory

PADMASRIAMBATI committed:
Create product-forms.spec.ts

PADMASRIAMBATI reviewed (commented):
Location Changed to tests/facility/settings/product/product-forms.spec.ts and removed unrelated change

yash-learner committed:
Merge branch 'develop' into issues/13930/redirect-after-edit

yash-learner commented on tests/facility/settings/product/product-forms.spec.ts:
```diff
@@ -0,0 +1,126 @@
+        page.getByRole("link", { name: /settings/i }).first(),
+      ).toBeVisible({ timeout: 10000 });
+    });
+    // --- END LOGIN ---
```
use 

```ts
test.use({ storageState: "tests/.auth/user.json" });
```
instead of redoing the login, this would pull the cred and log in the user automatically, so faster tests and this tests will focus what it needs to do

Go through other specs and  Inspire -> https://github.com/ohcnetwork/care_fe/blob/320c91eafd665a7b8418f4df42ec9e30a66b45d2/tests/facility/settings/product/products.spec.ts#L4

yash-learner commented on tests/facility/settings/product/product-forms.spec.ts:
```diff
@@ -0,0 +1,126 @@
+    await page.getByRole("button", { name: "Toggle Sidebar" }).click();
+    await page.getByRole("link", { name: /settings/i }).click();
+    await page.getByRole("link", { name: /products/i }).click();
+  });
```
You can directly  get the facility from `getFacilityId()` and construct the url `/facility/${facilityId}/settings/products`

Look  how its done in https://github.com/ohcnetwork/care_fe/blob/develop/tests/facility/settings/locations/location-creation.spec.ts

yash-learner commented on tests/facility/settings/product/product-forms.spec.ts:
```diff
@@ -0,0 +1,126 @@
+      productId = idMatch && idMatch[1] ? idMatch[1] : "unknown-product";
+
+      // Return to list to simulate starting the edit flow from the list/details page
+      await page.getByRole("link", { name: /products list/i }).click();
```
I think this create will not work

Because product creation happens via inventory

I think for your case its best and easy to edit an existing product without effecting anything else

yash-learner commented on tests/facility/settings/product/product-forms.spec.ts:
```diff
@@ -0,0 +1,126 @@
+      await page.getByRole("button", { name: "Edit" }).click();
+
+      // Update the product name
+      const nameInput = page.getByRole("textbox", { name: "Product Name" });
```
we can not edit name of the product in edit flow and I think there is no name field for product -> It gets the display from Product Knowledge and 
 
 You can edit following fields
 
 
<img width="1682" height="1364" alt="image" src="https://github.com/user-attachments/assets/e83bc485-7672-4755-8f30-190f00108b65" />

yash-learner commented on tests/facility/settings/product/product-forms.spec.ts:
```diff
@@ -0,0 +1,126 @@
+      );
+
+      // CRITICAL ASSERTION: Check that the URL is the correct details page
+      await expect(page).toHaveURL(expectedDetailsUrlRegex, { timeout: 10000 });
```
in most cases we do not need a explicit timeout because playwright automatically takes care

yash-learner reviewed (dismissed):
Good try on writing the specs 🚀

Here is some feedback 

Use camelCase for file names of specs - `productForm.spec.ts`

PADMASRIAMBATI committed:
Update product-forms.spec.ts

PADMASRIAMBATI committed:
Rename product-forms.spec.ts to productForm.spec.ts

PADMASRIAMBATI committed:
Refactor product form tests for robust navigation

Updated navigation method for stability and added critical checks for bug fixes.

yash-learner committed:
Merge branch 'develop' into pr/PADMASRIAMBATI/13951-1

yash-learner committed:
Update the spec and disable the form submit button when it is not dirty

nihal467 committed:
Merge branch 'develop' into issues/13930/redirect-after-edit

yash-learner committed:
Fix flakiness

yash-learner committed:
Merge branch 'issues/13930/redirect-after-edit' of https://github.com/padmasriambati/care_fe into pr/PADMASRIAMBATI/13951-1

yash-learner committed:
Merge branch 'develop' into issues/13930/redirect-after-edit

PADMASRIAMBATI commented:
All checks are passed

nihal467 commented:
LGTM
